test/integration/puppeteer.js

```diff
@@ -0,0 +1,156 @@
+  'webgl_worker_offscreencanvas',         // in a worker, not robust
+];
+
+// launch express server
+const app = express();
+app.use(express.static(__dirname + '/../..'));
+const server = app.listen(port, async () => {
+  try { await pup } catch (e) { console.error(e) } finally { server.close() }
+});
+
+// launch puppeteer with WebGL support in Linux
+let pup = puppeteer.launch({
+  headless: !process.env.VISIBLE,
+  args: [ '--use-gl=egl', '--no-sandbox' ]
+}).then(async browser => {
+
+  // prepare page
+  let pageSize, page = (await browser.pages())[0];
+  await page.setViewport({ width: 800, height: 600 });
+  const injection = fs.readFileSync('test/integration/deterministic-injection.js', 'utf8');
+  await page.evaluateOnNewDocument(injection);
+  page.on('console', msg => (msg.text().slice(0, 6) == 'Render') ? console.log(msg.text()) : {});
+  console.redLog = function(msg) { console.log(`\x1b[31m${msg}\x1b[37m`)}
+  console.greenLog = function(msg) { console.log(`\x1b[32m${msg}\x1b[37m`)}
+  page.on('response', async (response) => {
+    try { await response.buffer().then(buffer => { pageSize += buffer.length}); } catch {}
+  })
+
+  // find target files
+  const files = fs.readdirSync('./examples')
+    .filter(f => f.match(new RegExp(`.*\.(.html)`, 'ig')) && f != 'index.html' &&
+                (!process.env.FILE || (process.env.FILE && f == process.env.FILE + '.html')))
+    .map(s => s.slice(0, s.length - 5));
+
+  // forEach file with CI parallelism
+  let failedScreenshot = 0;
+  const isParallel = 'CI' in process.env;
+  const beginId = isParallel ? Math.floor(parseInt(process.env.CI.slice(0,1)) * files.length / 4) : 0;
+  const endId = isParallel ? Math.floor((parseInt(process.env.CI.slice(-1)) + 1) * files.length / 4) : files.length;
+  for (let id = beginId; id < endId; id++) {
+
+    // load target file
+    let file = files[id];
+    pageSize = 0;
+    if (exceptionList.includes(file)) continue;
+    try {
+      await page.goto(`http://localhost:${port}/examples/${file}.html`, { waitUntil: 'networkidle0', timeout: networkTimeout });
+    } catch (e) { /*console.log('Network timeout exceeded...');*/}
+
+    // prepare page
+    await page.evaluate(async (file, pageSize, minPageSize, maxPageSize, networkTax) => {
+      let button = document.getElementById('startButton');
+      if (button) button.click();
+      let style = document.createElement('style');
+      style.type = 'text/css';
+      style.innerHTML = `body { font size: 0 !important; }
+                         #info, button, input, body > div.dg.ac, body > div.lbl { display: none !important; }`;
+      document.getElementsByTagName('head')[0].appendChild(style);
+      let canvas = document.getElementsByTagName('canvas');
+      for (let i = 0; i < canvas.length; i++) {
+        if (canvas[i].height == 48) canvas[i].style.display = 'none';
+      }
+      if (file == 'misc_animation_authoring') {
```

Comment:
Interesting work! 👍 

Can you please explain why `misc_animation_authoring` needs a special handling? Isn't it possible to hide the UI by enhancing the above style definitions?

Reply:
No, but we can add a new class, for example `.lbl`, the same thing with stat.js. `if (canvas[i].height == 48) canvas[i].style.display = 'none';`

Reply:
I think it's okay to change a some example if it simplifies the code in your script 👍